Add PipeReader.Create from ReadOnlySequence<byte>

Alxandr opened the pull request:
Add new factory method to PipeReader to create a PipeReader from a ReadOnlySequence<byte>.

Fix #47990

Alxandr committed:
Add PipeReader.Create from ReadOnlySequence<byte>

Add new factory method to PipeReader to create a PipeReader from a
ReadOnlySequence<byte>.

Fix #47990

Alxandr commented:
I'm going to go through this again in a couple of days but I believe this is mostly done and ready for review.

@davidfowl said you wanted to take a look at it?

I've mostly copied tests from the stream pipe reader implementation, and removed tests that weren't applicable (like testing that the stream is disposed, etc.)

davidfowl commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,133 @@
+            _sequence = _sequence.Slice(consumed);
+        }
+
+        private CancellationTokenSource InternalTokenSource
```
This isn't needed.

Alxandr replied:
Per the documentation, you need to return cancelled on *next* request if the cancel has been called. I initially didn't have this but added it for that reason.

Alxandr replied:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs#L74

davidfowl replied:
I think it's fine for `CancelPendingRead` to noop if ReadAsync is always completed.

Alxandr replied:
That's a contract change according to the docs though? This is explicitly tested in the stream variant with TryRead. I didn't want to assume I could do that.

halter73 replied:
While CancelPendingRead does feel a bit useless given ReadAsync is always completed, not returning a canceled ReadResult from subsequent calls to ReadAsync/TryRead breaks the contract. I don't think we need a CancellationTokenSource. Using Interlocked.Exchange similar to how we do in Kestrel's [Http1UpgradeMessageBody](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/Http1UpgradeMessageBody.cs#L47) should be fine.

davidfowl replied:
I give you permission.

davidfowl replied:
In fact! I would update the docs to say that it may not fire if the ReadResult is already completed.

davidfowl replied:
@halter73 and I had some discussion offline. Here's how I feel about this. I think supporting cancellation is pointless but I can be convinced with a reasonable scenario. Also the current implementation is overkill as ReadAsync never waits. I'd love to understand how code could react to this, maybe I'm missing a simple test case

Alxandr replied:
Personally I don't have a dog in this fight (I believe that is the expression 😛). I added it because the API docs explicitly described this behaviour and there were unit tests that tested this behaviour, and I would consider not doing it a breaking change.

That being the case, having that breaking change (and announcing it I guess?) is perfectly fine for me.

@halter73 changing to atomics should be pretty straight forward. Not doing so was one of the reasons I created this PR in a draft state. If it's decided that the cancellation behaviour should remain as described in the docs, I will update the code to do that.

davidfowl replied:
I'll wait for the simplified version of cancellation

Alxandr replied:
I replaced it with some `Interlocked.Exchange` instructions and a single integer to keep track of whether or not cancellation has been requested. Let me know what you think :).

davidfowl commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,133 @@
+
+        private bool TryReadInternal(CancellationTokenSource source, out ReadResult result)
+        {
+            bool isCancellationRequested = source.IsCancellationRequested;
```
None of the cancellation logic is needed.

davidfowl commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,133 @@
+            return false;
+        }
+
+        private void ClearCancellationToken()
```
Same (not needed)

Alxandr committed:
Simplify cancel behaviour

Remove cancellation token source and replace with atomic int operations
to keep track of whether or not a cancellation has been requested.

Alxandr commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:
```diff
@@ -119,6 +119,16 @@ public static PipeReader Create(Stream stream, StreamPipeReaderOptions? readerOp
         }
 
+        /// <summary>
+        /// Creates a <see cref="PipeReader"/> wrapping the specified <see cref="ReadOnlySequence{T}"/>.
```
Is this fine? It points to the docs for `ReadOnlySequence<T>`, with no indication of it being only for bytes. Is this just how things are?

davidfowl replied:
Is that how the other docs are?

Alxandr replied:
I tried to google this issue (because you can't do `ReadOnlySequence<byte>`), and the idea here is that there is no docs page to link to for `ReadOnlySequence<byte>`, but the is for `ReadOnlySequence<T>`, so that is why I believe. I'm not sure where else I would find docs like this in .NET as I can't recall any off the top of my head at least.

Alxandr commented:
Apparently I don't know github well enough to know how to turn this PR into not-a-draft :P (or I might just not have access rights to do so). Anyways, I've gone though things again so it doesn't have to stay in draft state from my point of view.

davidfowl commented:
This looks good.

halter73 commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+                return new ValueTask<ReadResult>(result);
+            }
+
+            result = new ReadResult(_sequence, isCanceled: false, isCompleted: true);
```
Nit for clarity:

```suggestion
            result = new ReadResult(ReadOnlySequence<byte>.Empty, isCanceled: false, isCompleted: true);
```

Or maybe a comment pointing out that `_sequence` is always empty here.

Alxandr replied:
I agree. Added the requested changes :)

Alxandr committed:
Use @halter73's suggestion for clarity

Use `ReadOnlySequence<byte>.Empty` instead of `_sequence` when `_sequence` is known to be empty.

Co-authored-by: Stephen Halter <halter73@gmail.com>

davidfowl commented:
Arg, now there's a conflict. Sorry @Alxandr . Can you resolve it?

Alxandr commented:
Yep. Will do :)

stephentoub commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+
+namespace System.IO.Pipelines
+{
+    internal class SequencePipeReader : PipeReader
```
```suggestion
    internal sealed class SequencePipeReader : PipeReader
```

stephentoub commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+        {
+            ThrowIfCompleted();
+
+            // Fast path: did we consume everything?
+            if (consumed.Equals(_sequence.End))
+            {
+                _sequence = ReadOnlySequence<byte>.Empty;
+                return;
+            }
```
Is calling AdvanceTo on an already empty sequence the majority use case?  If not, I'd expect this to actually be a deoptimization.

halter73 replied:
This isn't checking if the sequence was already empty when the read started. This is checking if the entire ReadOnlySequence returned by ReadAsync was consumed which is very common.

stephentoub replied:
Ah. Ok.

stephentoub commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+        /// <inheritdoc />
+        public override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
```
How about putting this call in TryReadInternal rather than in both ReadAsync and TryRead?

Alxandr replied:
Done

stephentoub commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+            return TryReadInternal(out result);
+        }
+
+        private bool TryReadInternal(out ReadResult result)
```
Why is this method needed?  Why not just have all of this in TryRead and then have ReadAsync call TryRead?

Alxandr replied:
2 reasons. Firstly there used to be a bigger difference if I recall correctly (which isn't a reason to keep it), and second it mirrors the implementation for a stream pipe reader. That being said, as it is now I agree that it should probably just be removed, and I will do so.

stephentoub commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+
+        private bool TryReadInternal(out ReadResult result)
+        {
+            var isCancellationRequested = Interlocked.Exchange(ref _cancelNext, 0) == 1;
```
```suggestion
            bool isCancellationRequested = Interlocked.Exchange(ref _cancelNext, 0) == 1;
```

Alxandr replied:
Wait really? I thought it was recommended to use `var`?

davidfowl replied:
Not in dotnet/runtime

stephentoub commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+        {
+            ThrowIfCompleted();
+
+            if (TryReadInternal(out var result))
```
```suggestion
            if (TryReadInternal(out ReadResult result))
```

stephentoub commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs:
```diff
@@ -0,0 +1,104 @@
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            Interlocked.Exchange(ref _cancelNext, 1);
```
@davidfowl, I'm curious, why are the semantics of Cancel**Pending**Read that it cancels the next read if there isn't one pending?

davidfowl replied:
The next read will be cancelled.

stephentoub replied:
I understand.  I'm asking why it was designed that way :)

davidfowl replied:
CancelPendingRead is often used for waking up the loop in order to do something and this makes it easier to write code that doesn't need to do any other synchronization in order to make sure that the cancellation will be observed.

halter73 replied:
One concrete example of this is Kestrel's graceful shutdown. [We unset a `_keepAlive` flag and call CancelPendingRead](https://github.com/dotnet/aspnetcore/blob/a104f9d3c7990446326ca03d1dd777f41a6c4f53/src/Servers/Kestrel/Core/src/Internal/Http/Http1Connection.cs#L115-L124) on every connection's PipeReader. This guarantees that every connection observes the change to the `_keepAlive` flag even if the connection has not yet started reading the next request.

stephentoub replied:
That use makes sense.  I'd just have expected the most desirable semantic to be either a) cancel only the currently pending read or b) cancel the currently pending read and all future reads, but it's actually c) cancel the currently pending read or just the next read if one isn't currently pending.  I'd have though the graceful shutdown use case called out would have been better with (b), but in practice it probably doesn't matter.  I'd also have expected a different name than CancelPendingRead :smile:

halter73 replied:
The reason we don't want it to cancel all future reads is in case we're in the middle of reading a request. In that case, we observe the cancellation but essentially ignore it until we're done with the request.

stephentoub replied:
Thanks.  I spoke with @halter73 offline and he explained this to me in more detail.  I understand the intent now.

Alxandr committed:
Apply suggestions from code review

Replace `var` with `bool` and `ReadResult` in 2 places, and make `SequencePipeReader` sealed.

Co-authored-by: Stephen Toub <stoub@microsoft.com>

davidfowl commented:
@Alxandr still need to fix the merge conflict

Alxandr committed:
Merge branch 'main' into feat/pipe-reader-from-sequence

Alxandr commented:
> @Alxandr still need to fix the merge conflict

I know. I just had to be on a machine with a dev environment set up for that. The previous pushes has just been done through GitHub web UI.

Alxandr committed:
Remove TryReadInternal

Per suggestion of @stephentoub

davidfowl commented:
/azp run runtime-staging

davidfowl commented:
/azp run runtime

Alxandr commented:
I'm not sure whether or not these errors are related to code I did or not? I'm not seeing what went wrong in ADO (though, I'm probably just reading it wrong).

davidfowl commented:
Errors are unrelated.

davidfowl commented:
Thanks @Alxandr !
